### PR TITLE
Bilagird for splatfacto

### PR DESCRIPTION
**Pull request by KevinXu02**
Bilagird for splatfacto. The lr rate hasn't been tuned much, but this should work.

### CODE REVIEW

File: nerfstudio/models/splatfacto.py [hunk 1]
```diff
@@ -686,6 +700,27 @@ def _get_background_color(self):
...
+
+            out = slice(
+                bil_grids=self.bil_grids,
+                rgb=rgb,
+                xy=grid_xy,
+                grid_idx=torch.tensor(cam_idx, device=self.device, dtype=torch.long),
+            )
+            return out["rgb"]
+        else:
+            # do not apply bilateral grid to the first camera
```
Comment (jefequien) [inline nerfstudio/models/splatfacto.py:721]: Why is the bilateral grid not applied to the first camera?

Comment (KevinXu02) [reply nerfstudio/models/splatfacto.py:721]: I just let the first camera work as the reference, but this may hurt if the 1st photo is overexposured. I think I should make it also be applied.

File: nerfstudio/model_components/lib_bilagrid.py [hunk 2]
```diff
@@ -0,0 +1,478 @@
+# Copyright 2022 the Regents of the University of California, Nerfstudio Team and contributors. All rights reserved.
```
Comment (kerrj) [inline nerfstudio/model_components/lib_bilagrid.py:1]: We should remove this part about UCB and make it clear it's borrowed from the original paper authors (not written by us)

Comment (KevinXu02) [reply nerfstudio/model_components/lib_bilagrid.py]: fixed

File: nerfstudio/models/splatfacto.py [hunk 3]
```diff
@@ -184,6 +185,10 @@ class SplatfactoModelConfig(ModelConfig):
     """
     camera_optimizer: CameraOptimizerConfig = field(default_factory=lambda: CameraOptimizerConfig(mode="off"))
     """Config of the camera optimizer to use"""
+    use_bilateral_grid: bool = False
+    """If True, use bilateral grid to handle the ISP changes in the image space"""
```
Comment (kerrj) [inline nerfstudio/models/splatfacto.py:189]: can you add the citation here in the comment?

File: nerfstudio/models/splatfacto.py [hunk 4]
```diff
@@ -184,6 +185,10 @@ class SplatfactoModelConfig(ModelConfig):
...
+    gird_shape: Tuple[int, int, int] = (16, 16, 8)
```
Comment (kerrj) [inline nerfstudio/models/splatfacto.py:190]: typo `gird_shape`

File: nerfstudio/models/splatfacto.py [hunk 5]
```diff
@@ -686,6 +700,23 @@ def _get_background_color(self):
             raise ValueError(f"Unknown background color {self.config.background_color}")
         return background
 
+    def _apply_bilateral_grid(self, rgb: torch.Tensor, cam_idx: int, H: int, W: int) -> torch.Tensor:
+        # make xy grid
+        grid_y, grid_x = torch.meshgrid(
```
Comment (kerrj) [inline nerfstudio/models/splatfacto.py:705]: would be faster to create this meshgrid on the gpu and avoid the `.to(self.device)` line just below this

**Comment by kerrj** [comment]
Looks great! I think its ready to merge but if you have some time could you try evaluating how it affects PSNR on a few scenes or get some before/after renders? I think it's expected that the PSNR will drop since it's no longer trying to over-fit as heavily to each image.

File: nerfstudio/model_components/lib_bilagrid.py (hunk 2 above)
Comment (jb-ye) [reply nerfstudio/model_components/lib_bilagrid.py]: @KevinXu02 Just make sure there is a word "Copyright" in the python source code. For example, add a line saying:
```
# Copyright 2024 Yuehao Wang (https://github.com/yuehaowang)
```

**Comment by ichsan2895** [comment]
FYI
1. Feel free to use my render video & metrics in nerfstudio discord server for additional information. 
2. @jefequien said about the reason [why the metrics  of bilagird is lower than vanilla splatfacto](https://discord.com/channels/1025504971606724712/1026948651974660106/1264607432928137227)

**Comment by KevinXu02** [comment]
> > Scene (iters)	Metric	splatw(avg_app_embed)	bilags	splatfacto
> > bicycle (7000)	PSNR	22.73642	22.29577	22.96857
> > Time	473	330	260
> > kitchen (17000)	PSNR	30.37776	26.56442	30.76126
> > Time	1137	1241	851
> > room (17000)	PSNR	30.44047	28.42673	31.30814
> > Time	1095	1225	823
> 
> Should we implement the calculation of CC PSNR, CC SSIM, and CC LPIPS since the paper calculated that?

Yes! I'm going to adding it these days (but l may not got enough resources to rerun the experiments.

File: nerfstudio/model_components/lib_bilagrid.py [hunk 6]
```diff
@@ -0,0 +1,478 @@
...
+    """Returns total variation on multi-dimensional tensors.
+
+    Args:
+        x (torch.Tensor): The input tensor with shape $(B, C, ...)$, where $B$ is the batch size and $C$ is the channel size.
+    """
+    batch_size = x.shape[0]
+    tv = 0
+    for i in range(2, len(x.shape)):
+        n_res = x.shape[i]
+        idx1 = torch.arange(1, n_res).to(x.device)
```
Comment (kerrj) [inline nerfstudio/model_components/lib_bilagrid.py:93]: could we also change `.to(x.device)` to `arange(..., device=x.device)`?

File: nerfstudio/model_components/lib_bilagrid.py [hunk 7]
```diff
@@ -0,0 +1,547 @@
...
+
+import tensorly as tl
+import torch
+import torch.nn.functional as F
+from torch import nn
+
+tl.set_backend("pytorch")
+
+
+def color_correct(img: torch.Tensor, ref: torch.Tensor, num_iters: int = 5, eps: float = 0.5 / 255) -> torch.Tensor:
```
Comment (kerrj) [inline nerfstudio/model_components/lib_bilagrid.py:56]: this function was written by us right? would it make sense to add to one of the utils.py files? or does it make the most sense to keep in this file?

**Review by kerrj — APPROVED** [review]
I think this is ready to merge!
